pkg/core/blockchain.go

```diff
@@ -728,38 +729,82 @@ func (bc *Blockchain) resetStateInternal(height uint32, stage stateChangeStage)
 		// one stage, so persist periodically.
 		const persistBatchSize = 100 * headerBatchCount // count blocks only, should be enough to avoid OOM killer even for large blocks
 		var (
-			pBlocksStart                 = p
-			blocksCnt, batchCnt, keysCnt int
+			pBlocksStart        = p
+			blocksCnt, batchCnt int
+			keysCnt             uatomic.Int32
+			errCh               = make(chan error)
+			workers             uatomic.Int32
+			exit                = make(chan struct{})
 		)
-		for i := height + 1; i <= currHeight; i++ {
-			err := cache.DeleteBlock(bc.GetHeaderHash(int(i)))
-			if err != nil {
-				return fmt.Errorf("error while removing block %d: %w", i, err)
-			}
-			blocksCnt++
-			if blocksCnt == persistBatchSize {
-				keys, err = cache.Persist()
+		go func() {
+			for i := height + 1; i <= currHeight; i++ {
+				err = cache.DeleteBlock(bc.GetHeaderHash(int(i)))
 				if err != nil {
-					return fmt.Errorf("failed to persist intermediate batch removed blocks, transactions and AERs: %w", err)
+					// It's a critical error, stop the cycle after it.
+					err = fmt.Errorf("error while removing block %d: %w", i, err)
+					break
+				}
+				blocksCnt++
+				if blocksCnt == persistBatchSize {
+					workers.Inc()
+					blocksCnt = 0
+					batchCnt++
+
+					batchStart := p
+					batchID := batchCnt
+					go func() {
```

Comment:
Never ever start a new background persister before the old one has finished, it'll lead to OOM easily in some cases. Take a look at how `Run()` works.